File: 093_Course Schedule.cpp
// Course Schedule
// Determine if all courses can be finished given prerequisites (detect cycle in directed graph).
#include <bits/stdc++.h>
using namespace std;

bool canFinish(int numCourses, vector<vector<int>>& prerequisites){
    vector<vector<int>> g(numCourses);
    vector<int> indeg(numCourses,0);
    for(auto &p: prerequisites){ g[p[1]].push_back(p[0]); indeg[p[0]]++; }
    queue<int> q; for(int i=0;i<numCourses;++i) if(indeg[i]==0) q.push(i);
    int seen=0;
    while(!q.empty()){ int u=q.front(); q.pop(); ++seen; for(int v:g[u]) if(--indeg[v]==0) q.push(v); }
    return seen==numCourses;
}

int main(){
    int n=2; vector<vector<int>> pre={{1,0}};
    cout<<boolalpha<<"Can finish: "<<canFinish(n,pre)<<"\n";
    vector<vector<int>> bad={{1,0},{0,1}};
    cout<<"Can finish (cycle): "<<canFinish(2,bad)<<"\n";
    return 0;
}

// Time Complexity: O(N+E)
// Space Complexity: O(N+E)

File: 093_Course Schedule.py
"""
Course Schedule
Determine if you can finish all courses given prerequisites (topological check).
"""
from collections import deque, defaultdict
from typing import List

def can_finish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    g = defaultdict(list); indeg = [0]*numCourses
    for a,b in prerequisites:
        g[b].append(a); indeg[a]+=1
    q = deque([i for i in range(numCourses) if indeg[i]==0])
    seen=0
    while q:
        u = q.popleft(); seen+=1
        for v in g[u]:
            indeg[v]-=1
            if indeg[v]==0: q.append(v)
    return seen==numCourses

if __name__ == '__main__':
    print('Can finish:', can_finish(2, [[1,0]]))
    print('Can finish (cycle):', can_finish(2, [[1,0],[0,1]]))

# Time Complexity: O(N+E)
# Space Complexity: O(N+E)
